### PR TITLE
[set_containers] Add cifmw.general.set_containers module

### PR DESCRIPTION
**Pull request by arxcruz**
Replace the update_containers role invocation in deploy_architecture with the new cifmw.general.set_containers module. The module generates and optionally applies an OpenStackVersion CR, accepting a dynamic images list where each entry can override registry, org, name_prefix, and tag individually. This removes the rigid Jinja2 template in favor of a Python module that is easier to test, extend, and call from any playbook without role-level variable coupling.

Key design points:
- _OPENSTACK_SUFFIXES table enables partial overrides (e.g. tag-only) for any standard OpenStack image field without repeating the suffix.
- Empty name_prefix is supported so images that ship without the "openstack-" prefix (EDPM, IPA) can be built from parts rather than requiring a pre-assembled full_registry URL.
- backends list handles cinderVolumeImages / manilaShareImages nested dicts.
- Unit tests cover _build_url, _resolve_image, _resolve_backend, _build_cr, and the full run_module flow including apply, absent, and validation paths.

### CODE REVIEW

File: roles/cifmw_setup/tasks/deploy_architecture.yml [hunk 1]
```diff
@@ -209,13 +209,43 @@
...
+    # are migrated from cifmw_update_containers_* to cifmw_set_containers_*.
+    # Scenario-specific variables that also require explicit per-scenario migration:
+    #   cifmw_update_containers_cindervolumes  -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_manilashares   -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_watcher        -> cifmw_set_containers_extra_images
+    #   cifmw_update_containers_edpm_image_url -> cifmw_set_containers_images (name_prefix: "")
+    #   cifmw_update_containers_ipa_image_url  -> cifmw_set_containers_images (name_prefix: "")
+    registry: "{{ cifmw_set_containers_registry | default(cifmw_update_containers_registry) }}"
+    org: "{{ cifmw_set_containers_org | default(cifmw_update_containers_org) }}"
+    tag: "{{ cifmw_set_containers_tag | default(cifmw_update_containers_tag) }}"
```
Comment (rebtoor) [inline roles/cifmw_setup/tasks/deploy_architecture.yml]: If neither variable in the chain is defined, Ansible raises `UndefinedError`. The inner `default()` evaluates `cifmw_update_containers_registry` eagerly — when it is also undefined, it fails.

Safe pattern:

```yaml
registry: "{{ cifmw_set_containers_registry | default(cifmw_update_containers_registry | default(omit)) }}"
org: "{{ cifmw_set_containers_org | default(cifmw_update_containers_org | default(omit)) }}"
tag: "{{ cifmw_set_containers_tag | default(cifmw_update_containers_tag | default(omit)) }}"
```

(`include_openstack` and `apply` already chain `default(false)`, so they're fine.)

File: roles/cifmw_setup/tasks/deploy_architecture.yml [hunk 2]
```diff
@@ -209,13 +209,43 @@
...
+        )
+      }}
+    # TODO: remove the default() fallbacks below once all callers in ci-framework-jobs
+    # are migrated from cifmw_update_containers_* to cifmw_set_containers_*.
+    # Scenario-specific variables that also require explicit per-scenario migration:
+    #   cifmw_update_containers_cindervolumes  -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_manilashares   -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_watcher        -> cifmw_set_containers_extra_images
+    #   cifmw_update_containers_edpm_image_url -> cifmw_set_containers_images (name_prefix: "")
+    #   cifmw_update_containers_ipa_image_url  -> cifmw_set_containers_images (name_prefix: "")
```
Comment (rebtoor) [inline roles/cifmw_setup/tasks/deploy_architecture.yml:232]: **Regression risk during transition.** Cross-referencing `update_containers.j2` against downstream jobs, these old variables are actively set in production CI and have no backward-compat wiring here:

| Old variable | Where it's set | What silently breaks |
|---|---|---|
| `cifmw_update_containers_edpm_image_url` | `scenarios/uni/default-vars.yaml`, `scenarios/baremetal/default-vars.yaml`, periodic configs | `osContainerImage` missing from CR |
| `cifmw_update_containers_ipa_image_url` | same files | `ironicPythonAgentImage` missing |
| `cifmw_update_containers_barbican_custom_tag` | `zuul.d/adoption.yaml`, barbican component jobs | barbican images get wrong tag |
| `cifmw_update_containers_watcher` | uni03gamma, uni05epsilon, podified watcher scenarios | watcher images missing |
| `cifmw_update_containers_cindervolumes_extra` | uni09iota (Pure Storage FC) | extra cinder backends missing |


Consider adding backward-compat fallbacks in the `images` parameter construction (at least for the full-URL ones like `edpm_image_url` and `ipa_image_url`), or gate this change behind a feature flag so callers can opt in after migrating their variables.

File: roles/cifmw_setup/tasks/deploy_architecture.yml [hunk 3]
```diff
@@ -209,13 +209,43 @@
...
+      {{
+        cifmw_set_containers_apply |
+        default(cifmw_update_containers | default(false)) | bool
+      }}
+    images: >-
+      {{
+        (cifmw_set_containers_images | default([])) +
+        (cifmw_set_containers_extra_images | default([]))
+      }}
   tags:
```
Comment (rebtoor) [inline roles/cifmw_setup/tasks/deploy_architecture.yml:261]: Tag rename from `update_containers` to `set_containers` — one downstream hit confirmed:

```
# ci-framework-jobs/playbooks/adoption_uni_jobs/operators/install_operators.yaml
--skip-tags=openstack_ca,edpm_post,update_containers
```

After this merges, that `--skip-tags` will silently stop filtering this task. Worth a coordinated update or keeping both tags during transition.

File: plugins/modules/set_containers.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+    state = module.params["state"]
+    dest_path = module.params["dest_path"]
+    kubeconfig = module.params["kubeconfig"]
+    result = dict(changed=False, dest_path=dest_path)
+
+    if state == "absent":
+        if not os.path.exists(dest_path):
+            module.exit_json(**result)
+        if not module.check_mode:
+            if kubeconfig:
```
Comment (rebtoor) [inline plugins/modules/set_containers.py]: For `state=present`, cluster interaction requires `apply=true`. Here, `oc delete` fires whenever `kubeconfig` is set.

The old role used a **separate boolean** (`cifmw_update_containers_rollback`) to gate `oc delete` in `cleanup.yml` — not kubeconfig presence. A caller that sets `kubeconfig` for other purposes and passes `state=absent` expecting only local file cleanup will accidentally delete the CR from the cluster.

Suggest gating on `apply` for consistency:

```python
if kubeconfig and module.params["apply"]:
```

File: plugins/modules/set_containers.py [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+    return _build_url(reg, org, prefix, suffix, tag)
+
+
+def _resolve_backend(
+    backend, parent_url, mod_registry, mod_org, mod_prefix, mod_tag, module=None
+):
+    """Return (backend_name, image_url) for one item in a backends list."""
+    if isinstance(backend, str):
+        return backend, parent_url
+    name = backend.get("name")
```
Comment (rebtoor) [inline plugins/modules/set_containers.py:397]: `backend.get("name")` returns `None` when the key is missing, which silently inserts `None` as a dict key in the CR. Add validation:

```python
name = backend.get("name")
if not name:
    if module:
        module.fail_json(msg="backends entry is a dict but has no 'name' key")
    return None, parent_url
```

File: plugins/modules/set_containers.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+    "rabbitmqImage": "rabbitmq",
+    "swiftAccountImage": "swift-account",
+    "swiftContainerImage": "swift-container",
+    "swiftObjectImage": "swift-object",
+    "swiftProxyImage": "swift-proxy-server",
+    "testTempestImage": "tempest-all",
+}
+
+
+def _eff(per_image_val, module_val):
```
Comment (rebtoor) [inline plugins/modules/set_containers.py]: nit: `_eff` is cryptic for anyone reading the module cold. `_effective` or `_fallback` would be self-explanatory at no cost.

File: plugins/modules/set_containers.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+
+def _resolve_backend(
+    backend, parent_url, mod_registry, mod_org, mod_prefix, mod_tag, module=None
+):
+    """Return (backend_name, image_url) for one item in a backends list."""
+    if isinstance(backend, str):
+        return backend, parent_url
+    name = backend.get("name")
+    url = _resolve_image(backend, mod_registry, mod_org, mod_prefix, mod_tag, module)
+    return name, url if url is not None else parent_url
```
Comment (rebtoor) [inline plugins/modules/set_containers.py]: nit: Precedence is correct but explicit grouping prevents a double-take:

```python
return name, (url if url is not None else parent_url)
```

File: plugins/modules/set_containers.py [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+                ):
+                    return True
+    return False
+
+
+def _run_oc(module, args, kubeconfig):
+    oc_bin = module.get_bin_path("oc", required=True)
+    env = {"KUBECONFIG": kubeconfig} if kubeconfig else {}
+    rc, out, err = module.run_command(
+        [oc_bin] + args, environ_update=env if env else None
```
Comment (rebtoor) [inline plugins/modules/set_containers.py]: nit: The double conditional can be collapsed:

```python
environ_update = {"KUBECONFIG": kubeconfig} if kubeconfig else None
rc, out, err = module.run_command([oc_bin] + args, environ_update=environ_update)
```

File: tests/unit/modules/test_set_containers.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+            self._args(
+                include_openstack=True,
+                images=[dict(name="novaAPIImage", tag="special-tag")],
+            )
+        )
+        written = []
+
+        def fake_open(path, mode="r"):
+            m = mock_open()()
+            if mode == "r":
```
Comment (rebtoor) [inline tests/unit/modules/test_set_containers.py]: `written` and `fake_open` are defined but never used — the `with patch(...)` block below uses a different `mock_open()`.

Can we remove them?

**Review by rebtoor — CHANGES_REQUESTED** [review]

File: plugins/modules/set_containers.py [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+        if not os.path.exists(dest_path):
+            module.exit_json(**result)
+        if not module.check_mode:
+            if kubeconfig and module.params["apply"]:
+                rc, out, err = _run_oc(module, ["delete", "-f", dest_path], kubeconfig)
+                if rc != 0:
+                    module.fail_json(
+                        msg="oc delete failed", rc=rc, stdout=out, stderr=err
+                    )
+            os.remove(dest_path)
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py]: As the dest_path is a freeform input from the module parameters without sanitization this module can behave like universal delete and we only check that the path exists. We should probably encapsulate the os.remove in try-except block for various error conditions and fail the module gracefully should they occur.

File: plugins/modules/set_containers.py [hunk 11]
```diff
@@ -0,0 +1,587 @@
...
+                msg="images entry '{}': full_registry and container_suffix are mutually exclusive".format(
+                    spec["name"]
+                )
+            )
+
+    cr = _build_cr(module.params, module)
+    cr_yaml = yaml.dump(cr, default_flow_style=False, sort_keys=False)
+
+    existing = None
+    if os.path.exists(dest_path):
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py:597]: This does not guarantee That the dest_path is a file nor that we actually can read it (it might be for example existing folder returning true). os.path.isabs and os.path.is_file checks with graceful module failure could mitigate the risks here quite a bit.

File: plugins/modules/set_containers.py [hunk 12]
```diff
@@ -0,0 +1,587 @@
...
+                org=dict(type="str", default=None),
+                name_prefix=dict(type="str", default=None),
+                tag=dict(type="str", default=None),
+                backends=dict(type="list", elements="raw", default=None),
+            ),
+        ),
+        kubeconfig=dict(type="path", default=None),
+    )
+
+    module = AnsibleModule(argument_spec=module_args, supports_check_mode=True)
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py:530]: Using ansible.module_utils.common.arg_spec.ArgumentSpecValidator here could be very useful to ensure we actually have correct looking arguments provided before moving forward.

File: plugins/modules/set_containers.py [hunk 13]
```diff
@@ -0,0 +1,587 @@
...
+                )
+            )
+
+    cr = _build_cr(module.params, module)
+    cr_yaml = yaml.dump(cr, default_flow_style=False, sort_keys=False)
+
+    existing = None
+    if os.path.exists(dest_path):
+        with open(dest_path, "r") as fh:
+            existing = fh.read()
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py]: We probably should check here that the "existing" somewhat looks like CR, start by yaml parsing "is this file content actually yaml", etc. To make sure we don't overwrite some random file in the filesystem with the new CR.

File: plugins/modules/set_containers.py [hunk 14]
```diff
@@ -0,0 +1,587 @@
...
+
+    if cr_yaml != existing:
+        result["changed"] = True
+        if not module.check_mode:
+            dest_dir = os.path.dirname(dest_path)
+            if dest_dir:
+                os.makedirs(dest_dir, exist_ok=True)
+            with open(dest_path, "w") as fh:
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py]: Ditto with read. We have no guarantees at this point that the dest_path is a writeable file. We should try and fail gracefully.

File: tests/unit/modules/test_set_containers.py [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+    def test_absent_removes_existing_file(self):
+        set_module_args(self._args(state="absent"))
+        with patch(_MOD + ".os.path.exists", return_value=True), patch(
+            _MOD + ".os.remove"
+        ) as mock_rm:
+            with self.assertRaises(AnsibleExitJson) as cm:
+                set_containers.run_module()
+        mock_rm.assert_called_once_with("/tmp/set_containers.yml")
+        self.assertTrue(cm.exception.args[0]["changed"])
+
```
Comment (jokke-ilujo) [inline tests/unit/modules/test_set_containers.py:365]: Add tests for various invalid 'dest_path's (no write permissions, directory, empty string "", invalid content (say /etc/hosts for example))

**Review by jokke-ilujo — CHANGES_REQUESTED** [review]
I have some concerns of the dest_path file handling. Other than that looks quite good on the first pass.

File: plugins/modules/set_containers.py [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+                module.fail_json(
+                    msg="dest_path '{}' does not contain a YAML mapping".format(
+                        dest_path
+                    )
+                )
+        except yaml.YAMLError as exc:
+            module.fail_json(
+                msg="dest_path '{}' contains invalid YAML: {}".format(dest_path, exc)
+            )
+
```
Comment (jokke-ilujo) [inline plugins/modules/set_containers.py:615]: As it stands now should also except at least IsADirectoryError, PermissionError and IOError; or OSError if we don't want to be descriptive of the nature of the error of which caused the open() or read() to fail.

File: roles/cifmw_setup/tasks/deploy_architecture.yml [hunk 17]
```diff
@@ -209,13 +209,54 @@
...
+      }}
+    # TODO: remove the default() fallbacks below once all callers in ci-framework-jobs
+    # are migrated from cifmw_update_containers_* to cifmw_set_containers_*.
+    # Scenario-specific variables that also require explicit per-scenario migration:
+    #   cifmw_update_containers_cindervolumes  -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_manilashares   -> cifmw_set_containers_images (backends)
+    #   cifmw_update_containers_watcher        -> cifmw_set_containers_extra_images
+    #   cifmw_update_containers_edpm_image_url -> cifmw_set_containers_images (name_prefix: "")
+    #   cifmw_update_containers_ipa_image_url  -> cifmw_set_containers_images (name_prefix: "")
+    registry: "{{ cifmw_set_containers_registry | default(cifmw_update_containers_registry | default(omit)) }}"
```
Comment (evallesp) [inline roles/cifmw_setup/tasks/deploy_architecture.yml]: (blocking) possible error: I think this should be: {{ cifmw_set_containers_registry | default(cifmw_update_containers_registry) | default(omit) }}

**Review by evallesp — COMMENTED** [review]
LGTM in general!

File: plugins/modules/set_containers.py [hunk 18]
```diff
@@ -0,0 +1,633 @@
+#!/usr/bin/python
+
+# Copyright: (c) 2026, Red Hat
+# GNU General Public License v3.0+ (see COPYING or
```
Comment (holser) [inline plugins/modules/set_containers.py:4]: I guess it should be Apache Licence

**Review by holser — CHANGES_REQUESTED** [review]
